stl/inc/xutility

```diff
@@ -4299,69 +4299,111 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-struct _Ptr_cat_helper<_Elem, _Elem> { // determines _Ptr_cat's result when the types are the same
-    static constexpr bool _Really_trivial     = is_trivial_v<_Elem>;
-    static constexpr bool _Trivially_copyable = is_trivially_copyable_v<_Elem>;
+// VARIABLE TEMPLATE _Can_memop
+enum class _Memop_cat {
+    _Copy,
+    _Copy_uninitialized,
+    _Move,
+    _Move_uninitialized,
 };
 
-template <class _Anything>
-struct _Ptr_cat_helper<_Anything*, const _Anything*> {
-    // determines _Ptr_cat's result when all we do is add const to a pointer
-    static constexpr bool _Really_trivial     = true;
-    static constexpr bool _Trivially_copyable = true;
-};
+// Integral types are eligible for memcpy in very specific cases.
+// * Enumerations are treated as their underlying integral type
+// * They must be the same size. (`int == long` is eligible; `int != long long` isn't.)
+// * They are either both bool or both non bool
+template <class _Source, class _Dest>
+struct _Same_size_integral_or_bool
+    : bool_constant<sizeof(_Source) == sizeof(_Dest) && is_same_v<bool, _Source> == is_same_v<bool, _Dest>> {};
 
-template <class _Anything>
-struct _Ptr_cat_helper<_Anything*, volatile _Anything*> {
-    // determines _Ptr_cat's result when all we do is add volatile to a pointer
-    static constexpr bool _Really_trivial     = true;
-    static constexpr bool _Trivially_copyable = true;
-};
+template <class _Source, class _Dest>
+_INLINE_VAR constexpr bool _Can_memop_integrals =
+    conjunction_v<is_integral<_Source>, is_integral<_Dest>, _Same_size_integral_or_bool<_Source, _Dest>>;
 
-template <class _Anything>
-struct _Ptr_cat_helper<_Anything*, const volatile _Anything*> {
-    // determines _Ptr_cat's result when all we do is add cv to a pointer
-    static constexpr bool _Really_trivial     = true;
-    static constexpr bool _Trivially_copyable = true;
-};
+// Matching integer type are eligible for memcpy
+template <class _Source, class _Dest, _Memop_cat _Cat,
+    bool _Result = _Can_memop_integrals<_Unwrap_enum_t<_Source>, _Unwrap_enum_t<_Dest>>>
+_INLINE_VAR constexpr bool _Can_memop_elements = _Result;
 
-struct _False_copy_cat {
-    static constexpr bool _Really_trivial     = false;
-    static constexpr bool _Trivially_copyable = false;
-};
+// Pointer elements are eligible for memcpy when they point to the same type, ignoring cv-qualification.
+// This handles pointers to object types, pointers to void, and pointers to function types.
+// Performance note: This doesn't attempt to handle `object* != void*`.
+template <class _Source, class _Dest, _Memop_cat _Cat>
+_INLINE_VAR constexpr bool _Can_memop_elements<_Source*, _Dest*, _Cat, false> =
+    is_same_v<remove_cv_t<_Source>, remove_cv_t<_Dest>>;
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat : _False_copy_cat {}; // unwrap the pointer-like type and dispatch to _Ptr_cat_helper for copy
+// Types are eligible for memcpy / memmove when they are trivially copyable.

```

Comment:
What is the reasoning behind this? Types should be eligible for using `memcpy` / `memmove` when they are trivially copy constructible / move constructible / copy assignable / move assignable based on the operation. `copy` simply invokes copy assignment operator for every element so it should use `memmove` whenever the elements are trivially copy assignable. Similarly `move` simply invokes move assignment operator, `uninitialized_copy` simply invokes copy constructor and `uninitialized_move` simply invokes move constructor. `move` should not care about whether type is trivially copy assignable so `is_trivially_copyable` is too strict. Same with uninitialized versions which for some reason require type to be trivially default constructible (so types like `string_view` which is trivially copyable but not trivially default constructible won't be optimized).

Reply:
So this is just reformulating the existing code to use variable templates. I would rather not make functional changes within this PR if not needed.

I fully agree that we should have a more fine grained approach and this PR tries to build a flexible system so that we can more easily compose the requirements. This is essentially done by splitting it up into three independent parts

1. Are the iterators contiguous?
2. Does the type fulfill the necessary requirements?
3. Does the operation fulfill the necessary requirements?

Reply:
I understand. Still I would like a maintainer to answer my question.

Reply:
I believe the reasoning here is simply that this code is very old - the core of the optimization (detecting types that are safe to `memcpy/memmove`) was present in the codebase when I joined in 2007. My memories may be imperfect - this was one decade and two source control systems ago - but I seem to recall the original optimization being limited to only arithmetic and pointer types. I believe we then extended it to `is_pod` (which was extremely buggy at the time), and later that was reworked when `is_pod` became deprecated.

So the answer to the mystery of why we aren't using fine-grained type traits here is simply that they didn't exist back then. Also, we were confused by the fine-grained type traits for a while (there was a bug where `is_trivially_copy_constructible` was implemented with `is_trivially_copyable` - I missed all sorts of amazing bugs in the old days).

I think that separate issues/PRs to refine these optimizations to query the exact type traits needed would make a lot of sense. Thanks for bringing this up!

(I am having some success at clearing out my GitHub notifications but always feel free to ping us on Discord too if questions come up.)